### PR TITLE
Add support for NNC in the simulator

### PR DESCRIPTION
**Pull request by totto82**
1. NNC are added the grad, div and average operators
2. NNC are added the upwindSelector
3. NNC transmissibilities are added to the face transmissibilities

Tested on a modified SPE1 case where NNC are added.
Does not effect Norne. 

@qilicun I would appriciate if you could test this in combination with PINCH and MINPV

Depends on OPM/opm-parser#491

### CODE REVIEW

**Comment by qilicun** [comment]
@totto82 
Thanks for your effort, I will start the PINCH and MINPV work.

**Comment by qilicun** [comment]
When I build this patch, one error happened:
```
/project/eor/user/liuming/opm/opm-autodiff/opm/autodiff/AutoDiffHelpers.hpp:98: error: 'const class Dune::CpGrid' has no member named 'global_cell'
In file included from /project/eor/user/liuming/opm/opm-autodiff/opm/autodiff/BlackoilModelBase.hpp:29,
                 from /project/eor/user/liuming/opm/opm-autodiff/opm/autodiff/BlackoilModel.hpp:26,
                 from /project/eor/user/liuming/opm/opm-autodiff/opm/autodiff/SimulatorBase.hpp:30,
                 from /project/eor/user/liuming/opm/opm-autodiff/opm/autodiff/SimulatorFullyImplicitBlackoil.hpp:24,
                 from /project/eor/user/liuming/opm/opm-autodiff/examples/flow.cpp:47:
```

**Comment by alfbr** [comment]
Seems like a Dune version issue. @qilicun have you compiled with a fresh dune-cornerpoint build?

**Comment by GitPaean** [comment]
@qilicun 's error was reproduced here, with updated ert and all the opm related packages. 

The dune version is 2.2.1.

**Comment by qilicun** [comment]
@alfbr 

Yes, all the modules are up-to-date.

**Comment by bska** [comment]
> Seems like a Dune version issue

It is not.  This is a problem with the code.  The code references the property `global_cell` which does not exist in the `CpGrid` of module dune-cornerpoint.  As is, the code only compiles when targeting an opm-autodiff module that does not include bindings to dune-cornerpoint.  In other words, at present, this particular feature (PINCH, MINPV) cannot support `flow_cp`--meaning no parallelism.

**Comment by bska** [comment]
Oh, by the way, the fix to the build problems is to use @blattms's grid helpers.

**Comment by atgeirr** [comment]
The fix is as @bska commented, to not use UnstructuredGrid-specific interfaces.

At this point, development of the CpGrid-using (and parallel) version of Flow is of enough importance that I suggest we make dune-cornerpoint mandatory for opm-autodiff instead of optional. That would ensure we do not encounter this kind of problem anymore.

**Comment by totto82** [comment]
No it should run with CpGrid

**Comment by atgeirr** [comment]
I'm doing code review now, it compiles now for both flow and flow_cp, so that at least is ok.

File: opm/autodiff/AutoDiffHelpers.hpp [hunk 1]
```diff
@@ -127,11 +232,15 @@ struct HelperOps
             const IFIndex nif = h.internal_faces.size();
             typename ADFaceCellTraits<Grid>::Type
                 face_cells = faceCellsToEigen(g);
-            assert(nif == ifaceflux.size());
+
+            // num connections may possibly include NNCs
+            size_t num_nnc = h.nnc_trans.size();
+            size_t num_connections= nif + num_nnc;
+            assert(num_connections == ifaceflux.size());
```
Comment (atgeirr) [inline opm/autodiff/AutoDiffHelpers.hpp]: This line triggers a signed/unsigned comparison warning, as well as line 257. Eigen sizes are ints, so using `int` for num_connections and num_nnc is the simplest way to resolve this. Also use spaces around `=` in line 238.

File: opm/autodiff/AutoDiffHelpers.hpp [hunk 2]
```diff
@@ -58,6 +60,109 @@ struct HelperOps
...
+        int num_connections = num_internal;
+        int numNNC = 0;
+
+        // handle non-neighboring connections
+        std::shared_ptr<const NNC> nnc = eclState->getNNC();
+        if (nnc->hasNNC()) {
+            numNNC = nnc->numNNC();
+            num_connections += numNNC;
+            //std::cout << "Added " << numNNC << " NNC" <<std::endl;
+            nbi.resize(num_internal, 2);
```
Comment (atgeirr) [inline opm/autodiff/AutoDiffHelpers.hpp]: What is the purpose of this resize? Should it not already have this size?

**Comment by atgeirr** [comment]
The code looks correct. I do have one wish though: right now the new constructor duplicates a lot of the old constructor. That makes it harder to maintain and also harder to review. I would like you to fix this by getting rid of the old constructor. In order to enable all existing users to not pass an EclipseState you can add a default argument (`= EclipseStateConstPtr(nullptr)` should work), and tests in the constructor for the nullity of that argument. The simplest is probably to early on do:
```
std::shared_ptr<const NNC> nnc = eclState ? eclState->getNNC() : std::make_shared<const NNC>(nullptr);
const bool has_nnc = nnc && nnc->hasNNC();
```
With that change I will merge this.

**Comment by totto82** [comment]
@atgeirr 
Old constructor is removed.

**Comment by atgeirr** [comment]
There is another snag unfortunately: we need to update the polymer simulator as well. I will do that, and merge both PRs simultaneously.

**Pull request merged**